include/irr/asset/IVirtualTexture.h

```diff
@@ -140,46 +186,101 @@ class IVirtualTexture : public core::IReferenceCounted
     {
-        uint32_t sz = computeSquareSz(_img, _subres);
-        sz *= sz;
-        const uint32_t addr = core::morton2d_encode(_addr.pgTab_x, _addr.pgTab_y);
+        ISampler::SParams params;
+        params.AnisotropicFilter = m_tilePadding ? core::findMSB(m_tilePadding<<1) : 0u;
+        params.BorderColor = ISampler::ETBC_FLOAT_OPAQUE_WHITE;
+        params.CompareEnable = false;
+        params.CompareFunc = ISampler::ECO_NEVER;
+        params.LodBias = 0.f;
+        params.MaxLod = 0.f;
+        params.MinLod = 0.f;
+        params.MaxFilter = m_tilePadding ? ISampler::ETF_LINEAR : ISampler::ETF_NEAREST;
+        params.MinFilter = m_tilePadding ? ISampler::ETF_LINEAR : ISampler::ETF_NEAREST;
+        params.MipmapMode = ISampler::ESMM_NEAREST;
+        params.TextureWrapU = params.TextureWrapV = params.TextureWrapW = ISampler::ETC_CLAMP_TO_EDGE;
+
+        return params;
+    }
 
-        core::address_allocator_traits<pg_tab_addr_alctr_t>::multi_free_addr((*m_pageTableLayerAllocators)[_addr.pgTab_layer], 1u, &addr, &sz);
+    virtual core::smart_refctd_ptr<sampler_t> createSampler(const ISampler::SParams& _params) const = 0;
 
-        return true;
+    core::smart_refctd_ptr<sampler_t> getPhysicalPageSampler() const
+    {
+        if (!m_physicalPageSampler)
+            m_physicalPageSampler = createSampler(getPhysicalPageSamplerParams());
+        return m_physicalPageSampler;
+    }
+    core::smart_refctd_ptr<sampler_t> getPageTableSampler() const
+    {
+        if (!m_pageTableSampler)
+            m_pageTableSampler = createSampler(getPageTableSamplerParams());
+        return m_pageTableSampler;
     }
 
-    uint32_t countLevelsTakingAtLeastOnePage(const VkExtent3D& _extent, const ICPUImage::SSubresourceRange& _subres)
+    uint32_t getMaxAllocationPageCount() const
     {
-        const uint32_t baseMaxDim = core::roundUpToPoT(core::max(_extent.width, _extent.height))>>_subres.baseMipLevel;
+        auto pgtExtent = m_pageTable()->getCreationParameters().extent;
+        return pgtExtent.width*pgtExtent.height;
+    }
+
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t INVALID_SAMPLER_INDEX = 0xdeadbeefu;
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t INVALID_LAYER_INDEX = 0xdeadbeefu;
+
+    uint32_t findFreePageTableLayer() const
+    {
+        const uint32_t pgtLayers = m_pageTable->getCreationParameters().arrayLayers;
+        auto end = m_precomputed.layer_to_sampler_ix+pgtLayers;
+        auto found = std::find(m_precomputed.layer_to_sampler_ix, end, INVALID_SAMPLER_INDEX);
+        if (found == end)
+            return INVALID_LAYER_INDEX;
+        return found-m_precomputed.layer_to_sampler_ix;
+    }
+
+    uint32_t countLevelsTakingAtLeastOnePage(const VkExtent3D& _extent, uint32_t _baseLevel = 0u)
+    {
+        const uint32_t baseMaxDim = core::roundUpToPoT(core::max(_extent.width, _extent.height))>>_baseLevel;
         const int32_t lastFullMip = core::findMSB(baseMaxDim) - static_cast<int32_t>(m_pgSzxy_log2);
```

Comment:
this doesn't take into account that we round up the baseMaxDim to PoT (the rounding up of mip-levels)

it would need to be `core::findMSB(baseMaxDim-1u)+1` insead of  `core::findMSB(baseMaxDim`